### PR TITLE
NODE-2561 Attachment field for order

### PR DESCRIPTION
**Pull request by ivan-mashonskiy**

### CODE REVIEW

File: node/src/test/scala/com/wavesplatform/transaction/assets/exchange/OrderSpecification.scala [hunk 1]
```diff
@@ -246,8 +247,63 @@ class OrderSpecification extends PropSpec with ValidationMatcher with NTPTime {
...
+    createOrder(4, Some(ByteStr.fill(1)(1))).isValid(100).status shouldBe true
+
+    createOrder(4, Some(ByteStr.fill(Order.MaxAttachmentSize)(1))).isValid(100).status shouldBe true
+    createOrder(4, Some(ByteStr.fill(Order.MaxAttachmentSize + 1)(1)))
+      .isValid(100)
+      .hasError(s"attachment size should be <= ${Order.MaxAttachmentSize} bytes")
+    createOrder(4, Some(ByteStr.empty)).isValid(100).hasError("attachment size should be > 0")
+  }
+
+  property("NODE-966. Order V4 serialization roundtrip with non-existing attachment") {
```
Comment (vsuharnikov) [inline node/src/test/scala/com/wavesplatform/transaction/assets/exchange/OrderSpecification.scala:281]: Wondering, 
1. why do we need to test it without attachment?
2. do we need to test it **with** attachment? Old tests don't do this.

File: node/src/test/scala/com/wavesplatform/transaction/assets/exchange/EthOrderSpec.scala [hunk 2]
```diff
@@ -395,6 +395,82 @@ class EthOrderSpec
...
+    val orderWithAttachment    = testOrder(attachment)
+
+    val jsonWithoutAttachment = EthOrders.toEip712Json(orderWithoutAttachment)
+    checkOrderJson(jsonWithoutAttachment, 1, None)
+
+    val jsonWithAttachment = EthOrders.toEip712Json(orderWithAttachment)
+    checkOrderJson(jsonWithAttachment, 2, attachment)
+  }
+
+  it should "recover signer public key for order with and without attachment correctly (NODE-997)" in {
```
Comment (vsuharnikov) [inline node/src/test/scala/com/wavesplatform/transaction/assets/exchange/EthOrderSpec.scala:444]: Same question

File: node/src/test/scala/com/wavesplatform/state/diffs/smart/predef/TransactionBindingsTest.scala [hunk 3]
```diff
@@ -778,6 +780,58 @@ class TransactionBindingsTest extends PropSpec with PathMockFactory with EitherV
...
+         |match tx {
+         |  case o: ExchangeTransaction =>
+         |    let orderWithAttachment = o.buyOrder
+         |    let orderWithoutAttachment = o.sellOrder
+         |    orderWithAttachment.attachment == base58'$attachment' && orderWithoutAttachment.attachment == unit
+         |  case _ => true
+         |}
+       """.stripMargin
+
+    val compiledScript = TestCompiler(V8).compileExpression(script)
```
Comment (vsuharnikov) [inline node/src/test/scala/com/wavesplatform/state/diffs/smart/predef/TransactionBindingsTest.scala]: We won't forget to update this test for V9 in the future If we write:
```scala
VersionDic.all
  .filter(_.id >= V8.id)
  .map(TestCompiler(_).compileExpression(script))
  .foreach { compiledScript =>
    // code here
  }
```

**Review by vsuharnikov — APPROVED** [review]

**Review by yvaskov — APPROVED** [review]

File: node/src/test/scala/com/wavesplatform/state/BlockChallengeTest.scala [hunk 4]
```diff
@@ -463,16 +467,20 @@ class BlockChallengeTest extends PropSpec with WithDomain with ScalatestRouteTes
       d.rollbackTo(validOriginalBlock.header.reference)
 
       d.appendBlockE(challengingBlock) should beRight
-
-      // block snapshot contains only txs and block reward
+      val blockRewards = getLastBlockRewards(d)
+      // block diff contains only txs and block reward
```
Comment (xrtm000) [inline node/src/test/scala/com/wavesplatform/state/BlockChallengeTest.scala]: snapshot

File: node/src/test/scala/com/wavesplatform/transaction/smart/script/ScriptCompilerV1Test.scala [hunk 5]
```diff
@@ -26,8 +26,8 @@ class ScriptCompilerV1Test extends PropSpec with EitherValues with Inside {
   }
 
   property("fails on unsupported version") {
-    val script = scriptWithVersion("8".some)
-    ScriptCompiler(script, isAssetScript = false, estimator) shouldBe Left("Illegal directive value 8 for key STDLIB_VERSION")
+    val script = scriptWithVersion("9".some)
```
Comment (xrtm000) [inline node/src/test/scala/com/wavesplatform/transaction/smart/script/ScriptCompilerV1Test.scala]: `StdLibVersion.VersionDic.max.id + 1` is more versatile choice

File: node/src/test/scala/com/wavesplatform/state/BlockChallengeTest.scala (hunk 4 above)
Comment (ivan-mashonskiy) [reply node/src/test/scala/com/wavesplatform/state/BlockChallengeTest.scala]: Fixed

File: node/src/test/scala/com/wavesplatform/transaction/smart/script/ScriptCompilerV1Test.scala (hunk 5 above)
Comment (ivan-mashonskiy) [reply node/src/test/scala/com/wavesplatform/transaction/smart/script/ScriptCompilerV1Test.scala]: Fixed

**Review by xrtm000 — APPROVED** [review]

**Review by phearnot — APPROVED** [review]

**Pull request merged**